File: crates/ibex2/src/bindings/url.js
// URL and URLSearchParams.
//
// The CLASS SHAPE is here; every question about a URL is answered in Rust
// (stdlib/url.rs over the `url` crate — LLP 0059.000 §3.4: a partial URL
// parser is a security bug, not a missing feature). A URL object holds the
// eleven components the parser returned, and a setter asks Rust for the URL
// as it is after the assignment, so the setters' semantics are the spec's.
//
// A URLSearchParams holds its serialized query string and nothing else. Every
// method is one crossing that reads or rewrites that string in Rust, so the
// list semantics — order, duplicates, form-urlencoded escaping — have one
// implementation, and `url.searchParams` is a live view because it reads and
// writes `url.search` directly rather than keeping a copy.
(function (global) {
  "use strict";

  var parse = global.__ibex2_url_parse;
  var setField = global.__ibex2_url_set;
  var sp = {
    normalize: global.__ibex2_search_params_normalize,
    get: global.__ibex2_search_params_get,
    getAll: global.__ibex2_search_params_get_all,
    has: global.__ibex2_search_params_has,
    set: global.__ibex2_search_params_set,
    append: global.__ibex2_search_params_append,
    remove: global.__ibex2_search_params_delete,
    sort: global.__ibex2_search_params_sort,
    entries: global.__ibex2_search_params_entries,
  };

  var FIELDS = [
    "href", "origin", "protocol", "username", "password", "host", "hostname",
    "port", "pathname", "search", "hash",
  ];
  var SEARCH = FIELDS.indexOf("search");

  // Rust reports a parse failure as a message beginning "TypeError: "; the
  // spec wants an actual TypeError.
  function asTypeError(e) {
    var message = String(e && e.message !== undefined ? e.message : e);
    return new TypeError(message.replace(/^TypeError: /, ""));
  }

  var urls = new WeakMap(); // URL -> components array
  var views = new WeakMap(); // URL -> its searchParams, created once
  var params = new WeakMap(); // URLSearchParams -> { url: URL | null, query: string }

  function components(input, base) {
    try {
      return (base === undefined ? parse(String(input)) : parse(String(input), String(base))).split("\n");
    } catch (e) {
      throw asTypeError(e);
    }
  }

  function state(url) {
    var c = urls.get(url);
    if (!c) throw new TypeError("not a URL");
    return c;
  }

  function URL(input, base) {
    if (!(this instanceof URL)) throw new TypeError("URL must be constructed with new");
    if (arguments.length === 0) throw new TypeError("URL constructor requires a URL");
    urls.set(this, components(input, base));
  }

  FIELDS.forEach(function (field, index) {
    var descriptor = {
      enumerable: true,
      configurable: true,
      get: function () {
        return state(this)[index];
      },
    };
    if (field !== "origin") {
      descriptor.set = function (value) {
        var c = state(this);
        var next;
        try {
          next = setField(c[0], field, String(value));
        } catch (e) {
          // Only `href` throws; every other setter fails silently, and Rust
          // has already returned the URL unchanged for those.
          if (field === "href") throw asTypeError(e);
          return;
        }
        urls.set(this, next.split("\n"));
      };
    }
    Object.defineProperty(URL.prototype, field, descriptor);
  });

  Object.defineProperty(URL.prototype, "searchParams", {
    enumerable: true,
    configurable: true,
    get: function () {
      state(this);
      var view = views.get(this);
      if (!view) {
        view = new URLSearchParams();
        params.get(view).url = this;
        views.set(this, view);
      }
      return view;
    },
  });

  URL.prototype.toString = function () {
    return state(this)[0];
  };
  URL.prototype.toJSON = function () {
    return state(this)[0];
  };
  Object.defineProperty(URL.prototype, Symbol.toStringTag, { value: "URL", configurable: true });

  URL.canParse = function (input, base) {
    try {
      components(input, base);
      return true;
    } catch (e) {
      return false;
    }
  };
  URL.parse = function (input, base) {
    try {
      return new URL(input, base);
    } catch (e) {
      return null;
    }
  };

  // --- URLSearchParams ------------------------------------------------------

  function query(p) {
    if (p.url) return state(p.url)[SEARCH].replace(/^\?/, "");
    return p.query;
  }

  function write(p, next) {
    if (p.url) {
      // The spec's "update steps": the owner's search becomes the serialization,
      // and an empty one removes the `?` entirely.
      urls.set(p.url, setField(state(p.url)[0], "search", next).split("\n"));
    } else {
      p.query = next;
    }
  }

  function own(instance) {
    var p = params.get(instance);
    if (!p) throw new TypeError("not a URLSearchParams");
    return p;
  }

  function URLSearchParams(init) {
    if (!(this instanceof URLSearchParams)) {
      throw new TypeError("URLSearchParams must be constructed with new");
    }
    var p = { url: null, query: "" };
    params.set(this, p);
    if (init === undefined || init === null) return;
    if (typeof init === "string") {
      p.query = sp.normalize(init.charAt(0) === "?" ? init.slice(1) : init);
      return;
    }
    if (init instanceof URLSearchParams) {
      p.query = query(own(init));
      return;
    }
    if (typeof init !== "object") {
      p.query = sp.normalize(String(init));
      return;
    }
    if (typeof init[Symbol.iterator] === "function") {
      var pairs = Array.from(init);
      for (var i = 0; i < pairs.length; i++) {
        var pair = Array.from(pairs[i]);
        if (pair.length !== 2) {
          throw new TypeError("URLSearchParams init sequence entries must be name/value pairs");
        }
        p.query = sp.append(p.query, String(pair[0]), String(pair[1]));
      }
      return;
    }
    var names = Object.keys(init);
    for (var j = 0; j < names.length; j++) {
      p.query = sp.append(p.query, names[j], String(init[names[j]]));
    }
  }

  var proto = URLSearchParams.prototype;

  proto.get = function (name) {
    return sp.get(query(own(this)), String(name));
  };
  proto.getAll = function (name) {
    return JSON.parse(sp.getAll(query(own(this)), String(name)));
  };
  proto.has = function (name, value) {
    var q = query(own(this));
    return arguments.length > 1 && value !== undefined
      ? sp.has(q, String(name), String(value))
      : sp.has(q, String(name));
  };
  proto.set = function (name, value) {
    var p = own(this);
    write(p, sp.set(query(p), String(name), String(value)));
  };
  proto.append = function (name, value) {
    var p = own(this);
    write(p, sp.append(query(p), String(name), String(value)));
  };
  proto.delete = function (name, value) {
    var p = own(this);
    var q = query(p);
    write(
      p,
      arguments.length > 1 && value !== undefined
        ? sp.remove(q, String(name), String(value))
        : sp.remove(q, String(name))
    );
  };
  proto.sort = function () {
    var p = own(this);
    write(p, sp.sort(query(p)));
  };
  proto.toString = function () {
    return query(own(this));
  };
  Object.defineProperty(proto, "size", {
    configurable: true,
    get: function () {
      return JSON.parse(sp.entries(query(own(this)))).length;
    },
  });

  // Iteration hands out a snapshot of the pairs: one crossing, then the
  // engine's own array iterator.
  function pairs(instance) {
    return JSON.parse(sp.entries(query(own(instance))));
  }
  proto.entries = function () {
    return pairs(this)[Symbol.iterator]();
  };
  proto.keys = function () {
    return pairs(this)
      .map(function (pair) {
        return pair[0];
      })
      [Symbol.iterator]();
  };
  proto.values = function () {
    return pairs(this)
      .map(function (pair) {
        return pair[1];
      })
      [Symbol.iterator]();
  };
  proto[Symbol.iterator] = proto.entries;
  proto.forEach = function (callback, thisArg) {
    if (typeof callback !== "function") throw new TypeError("forEach requires a function");
    var list = pairs(this);
    for (var i = 0; i < list.length; i++) {
      callback.call(thisArg, list[i][1], list[i][0], this);
    }
  };
  Object.defineProperty(proto, Symbol.toStringTag, { value: "URLSearchParams", configurable: true });

  global.URL = URL;
  global.URLSearchParams = URLSearchParams;
})(globalThis);

File: crates/ibex2/tests/url.rs
//! `URL` and `URLSearchParams` through their binding: the shape is JavaScript,
//! every answer is Rust's (LLP 0059.000 §3.4). The parser itself is WPT-tested
//! in `wpt_url.rs`; this is the object on top of it.
#![cfg(feature = "hermes")]

use ibex2::engine::hermes::{DynamicCode, Hermes};

fn runtime() -> Hermes {
    let mut rt = Hermes::new(DynamicCode::Closed).expect("runtime");
    assert!(rt.install_stdlib());
    rt.install_bindings().expect("bindings");
    rt
}

fn eval(rt: &mut Hermes, program: &str) -> String {
    rt.eval(program)
        .unwrap_or_else(|e| panic!("{program}: {}", e.0))
}

#[test]
fn a_url_exposes_the_components_the_parser_normalized() {
    let mut rt = runtime();
    let out = eval(
        &mut rt,
        "(function () { const u = new URL('https://user:pw@Example.com:8080/a/../b?x=1#f'); \
         return [u.href, u.origin, u.protocol, u.username, u.password, u.host, u.hostname, \
         u.port, u.pathname, u.search, u.hash, String(u), JSON.stringify(u)].join('|'); })()",
    );
    assert_eq!(
        out,
        "https://user:pw@example.com:8080/b?x=1#f|https://example.com:8080|https:|user|pw|\
         example.com:8080|example.com|8080|/b|?x=1|#f|https://user:pw@example.com:8080/b?x=1#f|\
         \"https://user:pw@example.com:8080/b?x=1#f\""
    );
    // The default port is empty, and an absent search or hash is empty, not "?"/"#".
    assert_eq!(
        eval(&mut rt, "(function () { const u = new URL('https://h:443/'); return [u.port, u.search, u.hash, u.host].join('|'); })()"),
        "|||h"
    );
}

/// The case Exact's router runs on every navigation: a path against a base.
#[test]
fn a_relative_url_resolves_against_its_base() {
    let mut rt = runtime();
    assert_eq!(
        eval(
            &mut rt,
            "new URL('/settings', 'https://exact.local').pathname"
        ),
        "/settings"
    );
    assert_eq!(
        eval(&mut rt, "new URL('../c?q#h', 'https://h/a/b/').href"),
        "https://h/a/c?q#h"
    );
    assert_eq!(
        eval(&mut rt, "new URL('//other/x', 'https://h/').host"),
        "other"
    );
}

#[test]
fn an_invalid_url_is_a_type_error_and_can_parse_says_so_without_throwing() {
    let mut rt = runtime();
    assert_eq!(
        eval(&mut rt, "(function () { try { new URL('not a url'); return 'no throw'; } catch (e) { return e.constructor.name + ': ' + (e.message.indexOf('invalid URL') >= 0); } })()"),
        "TypeError: true"
    );
    assert_eq!(eval(&mut rt, "String(URL.canParse('not a url'))"), "false");
    assert_eq!(
        eval(&mut rt, "String(URL.canParse('/x', 'https://h'))"),
        "true"
    );
    assert_eq!(eval(&mut rt, "String(URL.parse('nope'))"), "null");
    assert_eq!(
        eval(&mut rt, "URL.parse('/x', 'https://h').href"),
        "https://h/x"
    );
    assert_eq!(
        eval(&mut rt, "(function () { try { URL('https://h'); return 'no throw'; } catch (e) { return e.constructor.name; } })()"),
        "TypeError"
    );
}

/// Setters are the spec's: they re-parse in Rust, fail silently, and only
/// `href` throws.
#[test]
fn setters_rewrite_the_url_with_the_specs_semantics() {
    let mut rt = runtime();
    let out = eval(
        &mut rt,
        "(function () { const u = new URL('https://h/p?a=1#x'); const log = []; \
         u.search = 'b=2'; log.push(u.href); \
         u.hash = ''; log.push(u.href); \
         u.pathname = 'q r'; log.push(u.pathname); \
         u.port = '8080'; log.push(u.host); \
         u.port = 'abc'; log.push(u.port); \
         u.host = 'h2:99'; log.push(u.hostname + ':' + u.port); \
         u.host = 'h3:80abc'; log.push(u.hostname + ':' + u.port); \
         u.host = 'h4:'; log.push(u.hostname + ':' + u.port); \
         u.protocol = 'http'; log.push(u.protocol); \
         u.protocol = 'mailto'; log.push(u.protocol); \
         u.protocol = 'https:garbage'; log.push(u.protocol); /* and 80 is now the default port, so it drops */ \
         u.username = 'me'; u.password = 'pw'; log.push(u.href); \
         u.href = 'https://z/'; log.push(u.origin); \
         try { u.href = 'nope'; log.push('no throw'); } catch (e) { log.push(e.constructor.name); } \
         log.push(u.href); \
         return log.join('|'); })()",
    );
    assert_eq!(
        out,
        "https://h/p?b=2#x|https://h/p?b=2|/q%20r|h:8080|8080|h2:99|h3:80|h4:80|http:|http:|https:|\
         https://me:pw@h4/q%20r?b=2|https://z|TypeError|https://z/"
    );
    assert_eq!(
        eval(
            &mut rt,
            "(function () { const u = new URL('https://h/'); u.origin = 'x'; return u.origin; })()"
        ),
        "https://h",
        "origin is read-only"
    );
}

/// `url.searchParams` is a live view over `url.search`, and the same object
/// every time.
#[test]
fn search_params_on_a_url_are_a_live_view() {
    let mut rt = runtime();
    let out = eval(
        &mut rt,
        "(function () { const u = new URL('https://h/?a=1&b=2'); const log = []; \
         u.searchParams.append('c', '3'); u.searchParams.delete('a'); log.push(u.href); \
         u.search = '?z=9'; log.push(u.searchParams.get('z')); \
         log.push(String(u.searchParams === u.searchParams)); \
         u.searchParams.set('q', 'a b&c'); log.push(u.search); \
         u.searchParams.delete('z'); u.searchParams.delete('q'); log.push(u.href); \
         return log.join('|'); })()",
    );
    assert_eq!(out, "https://h/?b=2&c=3|9|true|?z=9&q=a+b%26c|https://h/");
}

#[test]
fn search_params_stand_alone_with_every_construction_form_and_method() {
    let mut rt = runtime();
    let out = eval(
        &mut rt,
        "(function () { const log = []; \
         const p = new URLSearchParams('?b=2&a=1&a=3'); \
         p.sort(); log.push(String(p)); \
         log.push(p.getAll('a').join(',')); \
         log.push(String(p.has('a', '3')) + String(p.has('a', '2')) + String(p.has('zz'))); \
         p.delete('a', '1'); log.push(String(p)); \
         log.push([...p].map(e => e.join(':')).join(',')); \
         log.push([...p.keys()].join(',') + '/' + [...p.values()].join(',')); \
         log.push(String(p.size)); \
         const seen = []; p.forEach(function (v, k) { seen.push(k + '=' + v + this.tag); }, { tag: '!' }); log.push(seen.join(',')); \
         log.push(new URLSearchParams([['x', '1'], ['y', '2']]).toString()); \
         log.push(new URLSearchParams({ k: 'v w', n: 1 }).toString()); \
         log.push(new URLSearchParams(p).toString()); \
         log.push(new URLSearchParams('n=%E2%9C%93+x').get('n')); \
         log.push(String(new URLSearchParams().get('missing'))); \
         try { new URLSearchParams([['only-one']]); log.push('no throw'); } catch (e) { log.push(e.constructor.name); } \
         log.push(Object.prototype.toString.call(p)); \
         return log.join('|'); })()",
    );
    assert_eq!(
        out,
        "a=1&a=3&b=2|1,3|truefalsefalse|a=3&b=2|a:3,b:2|a,b/3,2|2|a=3!,b=2!|x=1&y=2|k=v+w&n=1|\
         a=3&b=2|✓ x|null|TypeError|[object URLSearchParams]"
    );
}
